### ISSUE TITLE
Oxidized Stop working - refs/heads/master' not found

### ISSUE THREAD

**Issue by jerebernard** [ytti/oxidized#1625]
Hello Everyone !!!

Oxidized Stop working since the update from the 8th of Nov.

This is the result when I start
```
oxidized    | reference 'refs/heads/master' not found
oxidized    | Puma starting in single mode...
oxidized    | * Version 3.12.0 (ruby 2.5.1-p57), codename: Llamas in Pajamas
oxidized    | * Min threads: 0, max threads: 16
oxidized    | * Environment: development
oxidized    | * Listening on tcp://0.0.0.0:8888
oxidized    | Use Ctrl-C to stop
oxidized    | reference 'refs/heads/master' not found
```
Oxidized start at the beginning then it come unreachable using a web browser with the log "reference 'refs/heads/master' not found"

Regards

**Comment by ytti** [comment]
Probably related to the refactor of output/git.rb 

No crash logs are created?

**Comment by jerebernard** [comment]
Yes there is a log crash : 

> 2018-11-12 08:19:16 UTC
> reference 'refs/heads/master' not found [Rugged::ReferenceError]
> --------------------------------------------------
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/output/git.rb:182:in `head'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/output/git.rb:182:in `update_repo'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/output/git.rb:170:in `update'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/output/git.rb:57:in `store'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/worker.rb:65:in `process_success'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/worker.rb:47:in `process'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/worker.rb:16:in `block in work'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/worker.rb:16:in `each'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/worker.rb:16:in `work'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/core.rb:37:in `run'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/core.rb:29:in `initialize'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/core.rb:4:in `new'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/core.rb:4:in `new'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/lib/oxidized/cli.rb:12:in `run'
> /var/lib/gems/2.5.0/gems/oxidized-0.24.0/bin/oxidized:8:in `<top (required)>'
> /usr/local/bin/oxidized:23:in `load'
> /usr/local/bin/oxidized:23:in `<main>'
> --------------------------------------------------

**Comment by ytti** [comment]
Is this fresh installation with no repo created?

What if in output/git.rb you change 

```ruby      
rescue Rugged::OSError, Rugged::RepositoryError => open_error
```

To

```ruby      
rescue Rugged::OSError, Rugged::RepositoryError, Rugged::ReferenceError => open_error
```

**Comment by jerebernard** [comment]
Installation is using docker-compose and the docker image oxidized/oxidized.
When I changed to your recommandation I got this

`first 'reference 'refs/heads/master' not found' was raised while opening git repo, then 'wrong argument type Rugged::Repository (expected String)' was while trying to create git repo`

In fact when the container is created with the git config 
```
output:
  default: git
  debug: false
  git:
    user: Oxidized
    email: backup@xxxx.fr
    single_repo: true
    repo: "/tmp/oxidized.git"
```
The git repo is created in the container but with the "/refs/heads/master" file is missing.

**Comment by ytti** [comment]
It looks like you did something else than what was asked. Can you verify the line you have is the line what was requested. IT looks like you broke ```Rugged::RepositoryError``` into ```Rugged::Repository```.

**Comment by jerebernard** [comment]
When I change output/git.rb 
```
rescue Rugged::OSError, Rugged::RepositoryError => open_error
To
rescue Rugged::OSError, Rugged::RepositoryError, Rugged::ReferenceError => open_error
```
I got this : `first 'reference 'refs/heads/master' not found' was raised while opening git repo, then 'wrong argument type Rugged::Repository (expected String)' was while trying to create git repo`

**Comment by ytti** [comment]
Hmm. Can you try to delete the repo first, and retry?

**Comment by jerebernard** [comment]
Yes for every test all the docker container  is deleted and recreated from the begining.

**Comment by ytti** [comment]
I can recreate the problem, it's only problem on initial repo creation, if you manually create it, you're fine. I'll have fix for you in moments.

**Comment by jerebernard** [comment]
I continue to troubleshoot and to compare with the config from 2 weeks ago and I found this.
When oxidized is load the git repo is initialized and it's this which doesn't work.
I can see tmp/oxidized.git but inside the config file is almost empty 
I can only see this
```
[core]
        bare = true
        repositoryformatversion = 0
        filemode = true
```
When I should see the git config
[core]
        bare = true
        repositoryformatversion = 0
        filemode = true
[user]
        name = Oxidized
        email = backup@xxx.fr
[remote "origin"]
        url = git@gitlab.xxx.fr:user/oxidized.git
        fetch = +refs/heads/*:refs/remotes/origin/*

**Comment by jerebernard** [comment]
It will be a pleasure to help and to test if needed.

**Comment by jerebernard** [comment]
Yes I just test and it works.
A huge thanks.

**Comment by ytti** [comment]
Thank you and sorry.

**Comment by jerebernard** [comment]
In fact the git output doesn't work fine as well. It deleted all existing files and recreate it.
But we lost all git history in order to do diff.

**Comment by ytti** [comment]
I'm not sure what happened, would need to look at the repo. It may be that the index is just written wrong. The size of the repo should be good indication if the data is still there.

I understood that you didn't have repo at all when you started oxidized, if you had repo already, then something is quite wrong with the code. Because the code thought you didn't have repo and created it.

**Comment by jerebernard** [comment]
In fact on our infra we have a gitlab where are located on backup directory.
Oxidized is base a docker container.
So when i launched my docker compose it create a new oxydized config and should download the git repo and update only device with config change
But now in fact the repo is downloaded and then all files are removed and recreated, so we lost the history.

**Comment by ytti** [comment]
The oxidized repo should have lived in a file outside t he container, the local repo was gone after you destroyed the container if it was inside the container.

But I mean you still have the repo in git lab.

So what you need to do is

a) stop oxidized
b) replace the local oxidized repo with cloned (bare) repo from gitlab
c) start oxidized

**Comment by ytti** [comment]
Your history should be just fine in gitlab. You must clone the gitlab repo, instead of letting oxidized create new repo.

**Comment by szpajder** [comment]
I came across the history breakage issue when importing my CVS repositories from Rancid (converted to Git repos with cvs2git). When I replaced the repo with a converted bare one an reran Oxidized, it has replaced all config files with the new ones, as if they didn't exist in the repo already.

The problem is that `update_repo` routine from git.rb updates the repo index, but bare repositories are not supposed to have one. So when `index.add` is run, it essentially creates a new empty index from scratch and then adds the file to it as if it was new, which breaks history.

The solution is to create the index manually. Assuming that the imported bare repository is in a directory named `imported-repo-bare.git`:

```
git clone imported-repo-bare.git imported-repo-nonbare
cp imported-repo-nonbare/.git/index imported-repo-bare.git
rm -rf imported-repo-nonbare
```

Now the `imported-repo-bare.git` can be moved to the Oxidized repo dir and Oxidized will update existing files properly without breaking history (assuming that file paths and names match, of course).
